### PR TITLE
Google Sign In/OpenID Connect Support

### PR DESCRIPTION
**Pull request by adamvy**
This PR add OpenID Connect support to nanos, and adds a "sign in with google" button to the SignIn view.

To use:
1. Set up oauth 2.0 with google by following the "App Setup" parts https://developers.google.com/identity/openid-connect/openid-connect#appsetup
2. Customize your AppConfig with the associated googleSignInClientId (you can do this on a per-theme basis)

### CODE REVIEW

File: src/foam/nanos/auth/JWTAuthService.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            
+            javax.json.JsonReader reader = javax.json.Json.createReader(new java.io.StringReader(header));
+            javax.json.JsonObject headerObject = reader.readObject();
+            reader.close();
+            
+            reader = javax.json.Json.createReader(new java.io.StringReader(body));
+            javax.json.JsonObject bodyObject = reader.readObject();
+            reader.close();
+            
+            // TODO validate JWT
```
Comment (adamvy) [inline src/foam/nanos/auth/JWTAuthService.js]: Need to implement this before merging.

File: src/foam/nanos/auth/UniqueUserService.js [hunk 2]
```diff
@@ -15,7 +15,7 @@ foam.INTERFACE({
       name: 'getUser',
       type: 'foam.nanos.auth.User',
       documentation: 'Helper logic function to reduce code duplication',
-      args: [ 'Context x', 'String identifier', 'String password' ]
+      args: [ 'Context x', 'String identifier' ]
```
Comment (adamvy) [inline src/foam/nanos/auth/UniqueUserService.js:18]: password was unused

File: src/foam/nanos/auth/UserAndGroupAuthService.js [hunk 3]
```diff
@@ -478,5 +417,83 @@ foam.CLASS({
         return true;
       `
     }
+  ],
+  static: [
+    {
+      name: 'loginUser',
```
Comment (adamvy) [inline src/foam/nanos/auth/UserAndGroupAuthService.js:423]: There's probably a better place for this, a standalone local only service to properly sign a session into a user, an authentication service for signing in, and the existing authentication service for doing permission checks.

But this works for now

File: src/foam/nanos/auth/services.jrl [hunk 4]
```diff
@@ -194,9 +197,11 @@ p({
   "name":"languageDAO",
   "lazy":false,
   "serve":true,
+  "authenticate": false,
```
Comment (adamvy) [inline src/foam/nanos/auth/services.jrl:200]: I'm not 100% on this change, but I don't see why languages is any different than countrydao/regiondao/etc

Comment (adamvy) [reply src/foam/nanos/auth/services.jrl:200]: in any case it was erroring out for my unauthed user due to unable to find languageDAO

**Comment by adamvy** [comment]
Google seems to rotate these keys faster than I expected, this code should be adjusted to just pull the JWKs from google directly on demand (and ideally cache them)

File: src/foam/nanos/auth/services.jrl (hunk 4 above)
Comment (jlhughes) [reply src/foam/nanos/auth/services.jrl:200]: @adamvy The languageDAO nspecs requires a GlobalReadAuthorizer. 
replace the authenticate:false with
       .setAuthorizer(new foam.nanos.auth.GlobalReadAuthorizer("language"))

File: src/foam/nanos/auth/AuthService.js [hunk 5]
```diff
@@ -66,6 +66,26 @@ foam.INTERFACE({
...
+    {
+      name: 'loginWithCredentials',
+      documentation: `
+        Log in the user with the provided credentials
+      `,
+      async: true,
+      type: 'foam.nanos.auth.User',
+      javaThrows: ['foam.nanos.auth.AuthenticationException'],
+      swiftThrows: true,
+      args: [
```
Comment (kgrgreer) [inline src/foam/nanos/auth/AuthService.js:78]: Could be just:
args: 'Context x, Object credentials',

File: src/foam/nanos/auth/Credentials.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+foam.CLASS({
```
Comment (kgrgreer) [inline src/foam/nanos/auth/Credentials.js:1]: Add header here and elsewhere, or leave it and we can do it.
/**
 * @license
 * Copyright 2024 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

File: src/foam/nanos/auth/JWK.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+});
+
+foam.CLASS({
+    package: 'foam.nanos.auth',
+    name: 'RSA256JWK',
+    extends: 'foam.nanos.auth.JWK',
+    properties: [
+        {
+            class: 'String',
+            name: 'n'
```
Comment (kgrgreer) [inline src/foam/nanos/auth/JWK.js:35]: I assume that this is what these are:
n	The modulus for the [RSA public key](https://tools.ietf.org/html/rfc7518#page-30).
e	The exponent for the [RSA public key](https://tools.ietf.org/html/rfc7518#page-30).

Would be good to add as comments or documentation:

File: src/foam/nanos/auth/User.js [hunk 8]
```diff
@@ -977,24 +977,23 @@ foam.CLASS({
...
-          throw new AuthenticationException("No session exists.");
+        
+        // check if user login enabled
+        if ( ! getLoginEnabled() ) {
+          throw new AccessDeniedException();
         }
-
-        if ( this instanceof LifecycleAware && ((LifecycleAware) this).getLifecycleState() != LifecycleState.ACTIVE ) {
-          throw new AuthenticationException("User is not active");
+
```
Comment (kgrgreer) [inline src/foam/nanos/auth/User.js:994]: Remove space characters.

File: src/foam/nanos/auth/UserAndGroupAuthService.js [hunk 9]
```diff
@@ -478,5 +417,83 @@ foam.CLASS({
...
+      documentation: `Helper function to reduce duplicated code.`,
+      type: 'User',
+      args: [
```
Comment (kgrgreer) [inline src/foam/nanos/auth/UserAndGroupAuthService.js:426]: Could be: args: 'Context serviceX, Context requestX, User user',

File: src/foam/nanos/auth/Credentials.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+            name: 'username'
+        },
+        {
+            class: 'String',
+            name: 'password'
+        }
+    ]
+});
+
+foam.CLASS({
```
Comment (kgrgreer) [inline src/foam/nanos/auth/Credentials.js:16]: We discourage having multiple models per file now, since doing so interferes with our new staging feature.

**Comment by adamvy** [comment]
closing in favo7ur of https://github.com/kgrgreer/foam3/pull/3550

**Comment by adamvy** [comment]
depreceated in favour of https://github.com/kgrgreer/foam3/pull/3550

**Pull request closed without merging**